src/satisfia/agents/makeMDPAgentSatisfia.py

```diff
@@ -444,9 +423,599 @@ def probability_add(p, key, weight):
+		lam = relativePosition(minAdmissibleQ(state, action), aleph4action, maxAdmissibleQ(state, action));
+		# (this is two numbers between 0 and 1.)
+		# use it to rescale aleph4action to the admissibility interval of the state that we landed in:
+		rescaledAleph4nextState = interpolate(minAdmissibleV(nextState), lam, maxAdmissibleV(nextState))
+		# (only this part preserves aspiration in expectation)
+		res = rescaledAleph4nextState # WAS: interpolate(steadfastAleph4nextState, rescaling4Successors, rescaledAleph4nextState)
+		if VERBOSE or DEBUG:
+			print(pad(state),"| | | propagateAspiration, state",prettyState(state),"action",action,"aleph4action",aleph4action,"Edel",Edel,"nextState",prettyState(nextState),":",res)
+		return res
+
+		""" Note on influence of Edel: 
+		It might seem that the (expected) delta received when taking action a in state s should occur
+		explicitly in some form in this formula, similar to how it occurred in the steadfast formula above.
+		This is not so, however. The expected delta is taken account of *implicitly* in the rescaling formula
+		via the use of min/maxAdmissibleQ(s,a) to compute lam but using min/maxAdmissibleV(s') in interpolating.
+		More precisely, one can prove that aspirations are kept in expectation. We want
+
+			aleph(s,a) = E(delta(s,a)) + E(aleph(s') | s'~(s,a)).
+
+		This can be shown to be true as follows:
+
+			min/maxAdmissibleQ(s,a) = E(delta(s,a)) + E(min/maxAdmissibleV(s') | s'~(s,a)),
+
+			lamdba = (aleph(s,a) - minAdmissibleQ(s,a)) / (maxAdmissibleQ(s,a) - minAdmissibleQ(s,a)),
+
+			rescaledAleph(s') = minAdmissibleV(s') + lambda * (maxAdmissibleV(s') - minAdmissibleV(s')),
+
+			E(delta(s,a)) + E(rescaledAleph(s') | s'~(s,a)) 
+			= E(delta(s,a)) + E(minAdmissibleV(s') | s'~(s,a)) 
+				+ lambda * (E(maxAdmissibleV(s') | s'~(s,a)) - E(minAdmissibleV(s') | s'~(s,a)))
+			= minAdmissibleQ(s,a) + lambda * (maxAdmissibleQ(s,a) - minAdmissibleQ(s,a))
+			= minAdmissibleQ(s,a) + (aleph(s,a) - minAdmissibleQ(s,a))
+			= aleph(s,a).
+
+		So the above rescaling formula is correctly taking account of received delta even without explicitly
+		including Edel in the formula.
+		"""
+
+	# Based on the policy, we can compute many resulting quantities of interest useful in assessing safety
+	# better than with the above myopic safety metrics. All of them satisfy Bellman-style equations:
+
+	# Actual Q and V functions of resulting policy (always returning scalars):
+	@lru_cache(maxsize=None)
+	def Q(self, state, action, aleph4action): # recursive
+		if DEBUG:
+			print("| Q", prettyState(state), action, aleph4action)
+
+		def sample():
+			Edel = expectedDelta(state, action)
+			if state.terminateAfterAction:
+				return Edel;
+			else:
+				nextState = self.world.transition(state, action)
+				nextAleph4state = self.propagateAspiration(state, action, aleph4action, Edel, nextState)
+				return Edel + V(nextState, nextAleph4state) # recursion
+		q = distribution.infer(sample).E()
```

Comment:
Hi again, Martin. Sorry to be insistent, but we must really avoid all sampling anywhere in our planning version of the code and rather compute the expectations *exactly* using loops. So in *all* these Bellmann equations. I did write the `expectation()` and `expected_reward()` helper functions in `WorldModel` for this. Use it something like this:
```
Edel = self.world.expected_reward([state], action, res2delta) 

def f(res):
    nextState, _, terminated, _, _, _ = res
    if terminated:
        return Edel;
    else:
        nextAleph4state = self.propagateAspiration(state, action, aleph4action, delta, nextState)
        return Edel + V(nextState, nextAleph4state) # recursion
q = self.world.expectation([state], action, f) 
```